Reader: send all posts across the redux bridge

samouri opened the pull request:
Part of the reduxification effort of #10984  and dependent on #22110.

Posts are pretty tricky and touch _all_ of Reader.  My thought is to tackle this in two layers:
1. Data Storage + requests to fetch posts. I plan on replacing the `PostStore.get` functionality.  Redux should be the only place to look for posts.  If a component needs a post, then it can use a query component that secretly uses flux under the hood.
2. (future PRs): Data fetching --> data layer

---
**To Test**
- run this brach locally. click around reader. You should see `state.reader.posts.items` populate with many posts! When you refresh it should be empty.
- open all of our streams, liking posts, searching for posts, recs, combined cards, everything.
- open up RelatedPosts and click into more RPs.  Especially for sites like Time and Uproxx
- use the `at` link and click an update pill. then click a gap

samouri commented on client/blocks/reader-combined-card/index.jsx:
```diff
@@ -137,4 +139,23 @@ class ReaderCombinedCard extends React.Component {
+	return postKey.postIds.map( postId => ( { feedId, blogId, postId } ) );
+}
+
+const getPostsByKeys = ( state, postKeys ) =>
```
note: this is inefficient. Ideally we can change the data structures involved here or make a memoized selector so that it doesn't cause needless rerenders

samouri commented:
dserve w/ update pill: https://dserve.a8c.com/?branch=update/reader/1.2-for-following&at=2018-01-01

samouri commented on client/lib/feed-post-store/index.js:
```diff
@@ -268,7 +180,8 @@ function receiveError( feedId, postId, error ) {
 
-	setPost( postId, {
+	setPost( {
+		global_ID: uuid(),
```
this line is icky. errors need to live in state, and we need to access it by postKey, but we don't have a real global_ID for them necessarily so I invent one here?

samouri commented:
The new changes to `/read/post/actions` should only affect related posts in the full post view and nothing else.

I noticed a `_should_reload` flag that stemmed from https://github.com/Automattic/wp-calypso/pull/8890 but could not figure out why or what they were used for. Especially considering that  these functions only apply to related-posts.  I've deleted the functionality and didn't notice any adverse effects.

blowery commented:
> I noticed a _should_reload flag that stemmed from #8890 but could not figure out why or what they were used for. Especially considering that these functions only apply to related-posts. I've deleted the functionality and didn't notice any adverse effects.

These are added by the API, not the client side. The code to refetch things marked as such was in `state/reader/posts/actions.js`

Pretty sure it's still needed to properly render related posts from VIPs and perhaps Jetpack sites.

samouri commented:
> Pretty sure it's still needed to properly render related posts from VIPs and perhaps Jetpack sites.

Why not regular posts from VIPs? For example in the following stream

blowery commented:
> Why not regular posts from VIPs? For example in the following stream

It's been a while since I had A Plan there, but iirc, The Plan was to reuse the redux loading bits to support that in the following stream too.

blowery commented:
Which is an obtuse way of saying "Yes, that would be nice"

blowery commented on client/state/reader/posts/selectors.js:
```diff
@@ -11,42 +17,30 @@ import { filter, keyBy } from 'lodash';
 
-	return postMapBySiteAndPost[ `${ siteId }-${ postId }` ];
-}
+export const getPostsByKeys = treeSelect(
```
why a separate treeSelect here? seems like the cache-key is overly specific and won't match much?

blowery replied:
just thinking about this vs reusing the `getPostMapByPostKey` and looking up via that.

blowery replied:
ah, because otherwise the returned array would change on every invocation and cause a re-render... hrm.

blowery replied:
maybe instead of resolving the post key to a post here, we add a new HoC that takes a postKey, resolves the key to a post, and then passes the post down to a wrapped component?

blowery replied:
then we can pass the array of postKeys, which doesn't change.

samouri replied:
I'm confused as to what the HoC would do exactly...

accept an array of postKeys or just an individual postKey?
How can we avoid creating a new array of posts at some level?

samouri replied:
ah. 

the right solution would be to refactor combined cards + recommended posts to not need the array of posts.  They should work solely off postKeys and then the individual subcomponents `CombinedCard` and `RelatedPosts` would be connected and would call `getPostForKey`

blowery commented on client/state/reader/posts/actions.js:
```diff
@@ -49,93 +27,69 @@ function fetchForKey( postKey ) {
+				ID: postKey.postId,
+				site_ID: postKey.blogId,
+				is_external: ! postKey.blogId,
+				global_ID: `${ postKey.feedId || 'na' }-${ postKey.blogId || 'na' }-${ postKey.postId }`,
```
this is inconsistent with the `uuid` earlier... do we need both?

samouri replied:
i should probably switch this one to a uuid as well

blowery commented on client/reader/stream/recommended-posts.jsx:
```diff
@@ -43,42 +45,32 @@ function handlePostClick( uiIndex, post ) {
-							<li className="reader-stream__recommended-posts-list-item" key={ post.global_ID }>
+							<li
+								className="reader-stream__recommended-posts-list-item"
+								key={ `${ index }-${ post && post.global_ID }` }
```
why are we rendering `null` posts here?

samouri replied:
I'll check

samouri replied:
because the `RelatedPost` is currently responsible for rendering the placeholders so it needs to go an extra layer without failing

blowery replied:
ahhhhh

samouri commented on client/state/reader/posts/actions.js:
```diff
@@ -49,93 +27,69 @@ function fetchForKey( postKey ) {
+	} );
 
+	// also save them after running the slow rules
+	Promise.all( normalizedPosts.map( runSlowRules ) ).then( processedPosts => {
```
good points. i'll fix it!

samouri replied:
It is significantly more performant to do a single batch update than to dispatch an update for each individually. 

thats a wrinkle

samouri replied:
I don't think the slow rules were even written to ever reject. `waitForImagesToLoad` only has a `resolve`

samouri replied:
I'm trying to manually trigger a reject within tests and I can't figure out how so far

blowery replied:
the only way I could think of reject getting triggered: something throws inside a rule

samouri replied:
easier said than done :).  the rules were written with tons of if checks and lots of lodash usage that make them fairly safe (at least towards my attempts at breaking it)

samouri commented:
ahh, so there is a current bug then in any location that uses VIPs like that.  Repro steps:

1. go to search
2. search for a "uproxx" and click one of their posts
3. you will only see an excerpt.  refresh and you'll see the whole thing

samouri commented:
Latest updates:
- removed more flux adapter nonsense form `stream/post.jsx`.
- removed flux nonsense from `FeaturedFeed` custom discover page header.  This needs testing as well.
- added the `_should_reload` bits back into place (VIP RPs work again)
- fixed both pill + updates.  they stopped working since removed flux nonsense since they all individually had a fetchPost if it couldn't be found.  They've been updated to various query components

samouri commented:
I've split out an issue for the VIP bug to https://github.com/Automattic/wp-calypso/issues/22299

jancavan commented:
Nothing appears broken to me 👍

blowery commented:
<img width="721" alt="screenshot 2018-02-09 16 52 13" src="https://user-images.githubusercontent.com/14350/36051908-ba64423e-0db9-11e8-802d-f88dfba9cdbf.png">

Missing author line...

jancavan commented:
@samouri I did notice something just now. When I try to share/reblog something, it's supposed to save my draft. When I exit the editor, I see this as a way to get back to my recently drafted post:

![screenshot 2018-02-09 13 56 53](https://user-images.githubusercontent.com/4924246/36052088-252fe744-0da1-11e8-9a49-c2351b126f15.png)

In this PR, I can just exit the Editor from reblogging something in Reader without my draft getting auto-saved.

blowery commented:
mmmm disregard mine. We have the same problem on prod. I also can't follow Design*Sponge or Spoon & Tamago. Might be odd VIPs?

/cc @benhuberman

samouri commented:
@blowery I do see some discrepancy on Discover in the authors between this branch and prod.
Notice the missing author name in this branch

**prod**
<img width="692" alt="screen shot 2018-02-09 at 5 06 44 pm" src="https://user-images.githubusercontent.com/4656974/36052446-b4a3e51e-0dbb-11e8-97aa-e303d24f7363.png">


**this branch**
<img width="700" alt="screen shot 2018-02-09 at 5 06 36 pm" src="https://user-images.githubusercontent.com/4656974/36052445-b3ad2080-0dbb-11e8-9f63-0f8ebbf90ba4.png">

jancavan commented:
@samouri I was also going to say that the number of Likes and Comments there are not the same.

jancavan commented:
There is also something weird going on with the Discover popover. It doesn't function the same way in prod.

If you try clicking on that Daily Prompt quick post button, the popover gets all wonky. It's making me lose track of where I am.

samouri commented:
Ok so the root of the issue was a dumb mistake in a connect call. within `post.jsx` I had written:
```jsx
getPostByKey( postKey )
```

instead of:
```jsx
getpostBykey( state, postKey )
```

---

I did make one more change on top of just fixing the bug.  I realized that the code was creating a new `discoverPick` object within each call of mapStateToProps which was causing rerenders.  The shape of `discoverPick`  was:

```
{ post, site }
```

I've instead separated it into two separate props: `discoverPost` and `discoverSite`

blowery commented on client/reader/stream/post.jsx:
```diff
@@ -24,7 +24,7 @@ class ReaderPostCardAdapter extends React.Component {
 	onClick = postToOpen => {
 		let referredPost;
-		if ( get( this.props, 'discoverPick.post' ) ) {
+		if ( get( this.props, 'discoverPost' ) ) {
```
this doesn't need to use `get`. Can just be `this.props.discoverPost`.

samouri replied:
theres a decent amount of cleanup that can still be done all around in terms of `lodash` usage.
We should probably fully get rid of the `PostCardAdapter` and put that functionality inside of `Post`

blowery commented:
Changes look good overall. Discover seems better. I can't repro @jancavan's bug, but maybe I'm doing it wrong. :)

benhuberman commented:
Just catching up with this issue @samouri @blowery -- I still see something odd in the Discover feed, though I don't actually know if that's a new thing or not: post picks display as expected, with an author name (where available) and site title -- both of them referring to the original, featured site, not to Discover or the editor who'd curated the post. Site picks, on the other hand, are shown as if they're "by" Discover and the editor who'd picked them. The different treatment feels odd to me, and I can't remember any conversation around this (though one could've very well taken place ages ago and I'd forgotten all about it, I suppose).

Any thoughts? It's not a crucial issue, but if we can have consistency, that would likely be ideal. I'll cc @clucasrowlands who might remember this conversation, if one took place, better than me.

clucasrowlands commented:
I don't recall a related conversation... but the attribution line (editor's name, Discover) next to a site pick was not there before, but I've also no idea if that's a new change or it's been like this for a while and we didn't notice!

(Agree with @benhuberman that it's not a crucial thing to fix, though it'd keep the feed consistent and less confusing.)

samouri commented:
I didn't intentionally change any behavior with this PR, it was meant just to be janitorial (moving from Flux-->Redux).

Here is a live link to right before this was merged: https://calypso.live/?hash=616a07b51828fdca415bf59f8af536cc479e9551
